### PR TITLE
[WIP] nan equality check for floating point datatypes

### PR DESCRIPTION
**Pull request by karthikeyann**
pandas compliant equality check for nan for floating point datatypes
closes https://github.com/rapidsai/cudf/issues/2191

### CODE REVIEW

**Comment by karthikeyann** [comment]
This code change will slowdown all row comparison operations.

File: cpp/src/table/legacy/device_table_row_operators.cuh [hunk 1]
```diff
@@ -54,6 +56,41 @@ struct elements_are_equal {
...
+  template <typename ColumnType>
+  __device__ __forceinline__ 
+    typename std::enable_if_t<std::is_floating_point<ColumnType>::value, bool>
+    operator()(gdf_column const& lhs,
+                                             gdf_size_type lhs_index,
+                                             gdf_column const& rhs,
+                                             gdf_size_type rhs_index,
+                                             bool nulls_are_equal = false) {
+    if (nullable) {
+      bool const lhs_is_valid{gdf_is_valid(lhs.valid, lhs_index)};
```
Comment (harrism) [inline cpp/src/table/legacy/device_table_row_operators.cuh]: Does `nulls_are_equal` imply `NULL == NaN` is `true`? If so, I think I would restructure this a bit:

```
const ColumnType lhs_data = static_cast<ColumnType const*>(lhs.data)[lhs_index];
const ColumnType rhs_data = static_cast<ColumnType const*>(rhs.data)[rhs_index];
bool lhs_is_valid{not isnan(lhs_data)};
bool rhs_is_valid{not isnan(rhs_data)};

if (nullable) {
    lhs_is_valid ||= gdf_is_valid(lhs.valid, lhs_index);
    rhs_is_valid ||= gdf_is_valid(rhs.valid, rhs_index);
}

// If both values are invalid
if (not lhs_is_valid and not rhs_is_valid) {
    return nulls_are_equal;
}

// If only one value is null, they can never be equal
if (lhs_is_valid != rhs_is_valid) {
    return false;
}
   
return lhs_data == rhs_data;
```

File: cpp/src/table/legacy/device_table_row_operators.cuh [hunk 2]
```diff
@@ -187,6 +224,46 @@ struct typed_row_inequality_comparator
...
+    const ColType lhs_data = static_cast<const ColType *>(lhs_column->data)[lhs_row];
+    const ColType rhs_data = static_cast<const ColType *>(rhs_column->data)[rhs_row];
+
+    bool isValid1, isValid2;
+    if (nullable)
+    {
+      isValid1 = gdf_is_valid(lhs_column->valid, lhs_row);
+      isValid2 = gdf_is_valid(rhs_column->valid, rhs_row);
+    }
+    else
```
Comment (harrism) [inline cpp/src/table/legacy/device_table_row_operators.cuh]: Is this logic correct? You only check for NaNs if it's not nullable? What if you have a nullable column with non-null NaN elements?

I think we need to decide whether NaN == NULL is true or false.

File: cpp/src/table/legacy/device_table_row_operators.cuh [hunk 3]
```diff
@@ -187,6 +224,46 @@ struct typed_row_inequality_comparator
         return State::False;
     }
   }
+  template <typename ColType>
+  __device__
+    typename std::enable_if_t<std::is_floating_point<ColType>::value, State>
+      operator()(gdf_index_type lhs_row, gdf_index_type rhs_row,
+                 gdf_column const *lhs_column, gdf_column const *rhs_column,
```
Comment (harrism) [inline cpp/src/table/legacy/device_table_row_operators.cuh]: Why does this operator take pointers to columns and the other ones take references?

File: cpp/src/table/legacy/device_table_row_operators.cuh (hunk 2 above)
Comment (kkraus14) [reply cpp/src/table/legacy/device_table_row_operators.cuh]: That should absolutely be false.

Comment (karthikeyann) [reply cpp/src/table/legacy/device_table_row_operators.cuh]: @harrism Yes. I figured later that I missed this case. I will fix it now.
`NaN == NULL` is `false` - Is this documented somewhere?

**Comment by karthikeyann** [comment]
> I'd be interested to know what the cost of this change is. Also have some concerns about the logic / structure. See comments.
Benchmarked here https://github.com/rapidsai/cudf/issues/2191#issuecomment-529330687

**Comment by harrism** [comment]
Move to 0.12

**Comment by jrhemstad** [comment]
fyi, this is obviated by the new libcudf++ comparators which already handle special casing for floating point. This is only relevant for legacy APIs.

**Comment by harrism** [comment]
Ah, good to know. Will leave open for now but hopefully we can close it soon.

**Comment by karthikeyann** [comment]
Can I close this PR?

**Pull request closed without merging**